File: blog-backend/app/routers/tags.py
from fastapi import APIRouter, Depends, HTTPException, status, Query, Path
from sqlalchemy.orm import Session
from typing import List, Optional
from sqlalchemy import func

from app import models
from app.core import security
from app.core.database import get_db
from app.core.cache import clear_cache_by_prefix
from app.schemas.tag import TagCreate, TagResponse, TagUpdate, TagWithCount, TagBatchDeleteRequest
from app.schemas.article import ArticleList
from app.services.tag_service import TagService

router = APIRouter(prefix="/tags", tags=["tags"])

@router.post("", response_model=TagResponse, status_code=status.HTTP_201_CREATED)
async def create_tag(
    tag: TagCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Create a new tag.
    """
    return TagService.create_tag(db=db, tag=tag)

@router.get("", response_model=List[TagWithCount])
async def get_tags(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """
    Get all tags with article count.
    """
    return TagService.get_tags_with_count(db=db, skip=skip, limit=limit)

@router.get("/{tag_id}", response_model=TagResponse)
async def get_tag(
    tag_id: int,
    db: Session = Depends(get_db)
):
    """
    Get a specific tag by ID.
    """
    tag = TagService.get_tag_by_id(db=db, tag_id=tag_id)
    if tag is None:
        raise HTTPException(status_code=404, detail="Tag not found")
    return tag

@router.put("/{tag_id}", response_model=TagResponse)
async def update_tag(
    tag_id: int,
    tag: TagUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Update a tag.
    """
    return TagService.update_tag(db=db, tag_id=tag_id, tag_update=tag)

@router.delete("/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_tag(
    tag_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Delete a tag.
    """
    success = TagService.delete_tag(db=db, tag_id=tag_id)
    if not success:
        raise HTTPException(status_code=404, detail="Tag not found")
    return None

@router.post("/batch-delete", status_code=status.HTTP_200_OK)
async def batch_delete_tags(
    request: TagBatchDeleteRequest,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    批量删除标签。

    参数:
    - **tag_ids**: 要删除的标签ID列表

    返回:
    - 200 OK: 删除成功，返回删除的数量
    - 403 Forbidden: 没有权限删除
    """
    # 检查用户是否为管理员
    if current_user.role != "admin":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="只有管理员可以批量删除标签"
        )

    if not request.tag_ids:
        return {"deleted_count": 0, "message": "没有指定要删除的标签"}

    # 查询并删除标签
    deleted_count = 0
    for tag_id in request.tag_ids:
        success = TagService.delete_tag(db=db, tag_id=tag_id)
        if success:
            deleted_count += 1

    # 清除相关缓存
    clear_cache_by_prefix("tags_")

    return {
        "deleted_count": deleted_count,
        "message": f"成功删除 {deleted_count} 个标签"
    }

@router.get("/by-name/{name}", response_model=TagResponse)
async def get_tag_by_name(
    name: str,
    db: Session = Depends(get_db)
):
    """
    Get a specific tag by name.
    """
    tag = TagService.get_tag_by_name(db=db, name=name)
    if tag is None:
        raise HTTPException(status_code=404, detail="Tag not found")
    return tag

@router.get("/{tag_id}/articles", response_model=List[ArticleList])
async def get_articles_by_tag(
    tag_id: int = Path(..., title="The ID of the tag to get articles for"),
    skip: int = Query(0, ge=0),
    limit: int = Query(10, ge=1, le=100),
    db: Session = Depends(get_db)
):
    """
    Get all articles with a specific tag.
    """
    return TagService.get_articles_by_tag(db=db, tag_id=tag_id, skip=skip, limit=limit)

@router.post("/articles/{article_id}/tags")
async def add_tag_to_article(
    article_id: int = Path(..., title="The ID of the article to add the tag to"),
    tag_id: int = Query(..., title="The ID of the tag to add"),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Add a tag to an article.
    """
    return {"success": TagService.add_tag_to_article(db=db, article_id=article_id, tag_id=tag_id)}

@router.delete("/articles/{article_id}/tags/{tag_id}")
async def remove_tag_from_article(
    article_id: int = Path(..., title="The ID of the article to remove the tag from"),
    tag_id: int = Path(..., title="The ID of the tag to remove"),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Remove a tag from an article.
    """
    return {"success": TagService.remove_tag_from_article(db=db, article_id=article_id, tag_id=tag_id)}

@router.put("/articles/{article_id}/tags")
async def update_article_tags(
    article_id: int = Path(..., title="The ID of the article to update tags for"),
    tag_ids: List[int] = Query(..., title="The IDs of the tags to set"),
    db: Session = Depends(get_db),
    current_user: models.User = Depends(security.get_current_user)
):
    """
    Update all tags for an article.
    """
    tags = TagService.update_article_tags(db=db, article_id=article_id, tag_ids=tag_ids)
    return {"tags": [tag.name for tag in tags]}

@router.get("/articles/{article_id}/tags", response_model=List[TagResponse])
async def get_article_tags(
    article_id: int = Path(..., title="The ID of the article to get tags for"),
    db: Session = Depends(get_db)
):
    """
    Get all tags for an article.
    """
    return TagService.get_article_tags(db=db, article_id=article_id)

File: blog-frontend/src/api/modules/tags.js
/**
 * 标签相关 API
 * 封装所有与标签数据相关的网络请求
 */

import apiClient from '../core/apiClient';
import { API_PATHS } from '../core/apiPaths';
import { getAuthHeaders } from '../core/apiClient';

/**
 * 获取所有标签及关联文章数量
 * @param {Object} params - 查询参数
 * @param {number} [params.skip=0] - 跳过的标签数量
 * @param {number} [params.limit=100] - 返回的标签数量
 * @returns {Promise} 返回包含标签列表的Promise对象
 */
export const getTags = (params = {}) => {
  return apiClient.get(API_PATHS.TAGS.BASE, {
    params: {
      skip: params.skip || 0,
      limit: params.limit || 100
    }
  });
};

/**
 * 根据ID获取标签
 * @param {number} id - 标签ID
 * @returns {Promise} 返回包含标签数据的Promise对象
 */
export const getTagById = (id) => {
  return apiClient.get(API_PATHS.TAGS.BY_ID(id));
};

/**
 * 根据名称获取标签
 * @param {string} name - 标签名称
 * @returns {Promise} 返回包含标签数据的Promise对象
 */
export const getTagByName = (name) => {
  return apiClient.get(API_PATHS.TAGS.BY_NAME(name));
};

/**
 * 创建新标签
 * @param {Object} tagData - 标签数据
 * @param {string} tagData.name - 标签名称
 * @returns {Promise} 返回创建的标签数据
 */
export const createTag = (tagData) => {
  return apiClient.post(API_PATHS.TAGS.BASE, tagData, {
    headers: getAuthHeaders()
  });
};

/**
 * 更新标签
 * @param {number} id - 标签ID
 * @param {Object} tagData - 标签数据
 * @param {string} tagData.name - 标签名称
 * @returns {Promise} 返回更新的标签数据
 */
export const updateTag = (id, tagData) => {
  return apiClient.put(API_PATHS.TAGS.BY_ID(id), tagData, {
    headers: getAuthHeaders()
  });
};

/**
 * 删除标签
 * @param {number} id - 标签ID
 * @returns {Promise} 返回删除结果
 */
export const deleteTag = (id) => {
  return apiClient.delete(API_PATHS.TAGS.BY_ID(id), {
    headers: getAuthHeaders()
  });
};

/**
 * 批量删除标签
 * @param {Array<number>} tagIds - 标签ID数组
 * @returns {Promise} 返回删除结果
 */
export const batchDeleteTags = (tagIds) => {
  return apiClient.post(API_PATHS.TAGS.BATCH_DELETE, {
    tag_ids: tagIds
  }, {
    headers: getAuthHeaders()
  });
};

/**
 * 获取具有特定标签的所有文章
 * @param {number} tagId - 标签ID
 * @param {Object} params - 查询参数
 * @param {number} [params.skip=0] - 跳过的文章数量
 * @param {number} [params.limit=10] - 返回的文章数量
 * @returns {Promise} 返回包含文章列表的Promise对象
 */
export const getArticlesByTag = (tagId, params = {}) => {
  return apiClient.get(API_PATHS.TAGS.ARTICLES(tagId), {
    params: {
      skip: params.skip || 0,
      limit: params.limit || 10
    }
  });
};

/**
 * 给文章添加标签
 * @param {number} articleId - 文章ID
 * @param {number} tagId - 标签ID
 * @returns {Promise} 返回添加结果
 */
export const addTagToArticle = (articleId, tagId) => {
  return apiClient.post(API_PATHS.TAGS.ADD_TO_ARTICLE(articleId), null, {
    params: { tag_id: tagId },
    headers: getAuthHeaders()
  });
};

/**
 * 更新文章的所有标签
 * @param {number} articleId - 文章ID
 * @param {Array<number>} tagIds - 标签ID数组
 * @returns {Promise} 返回更新结果
 */
export const updateArticleTags = (articleId, tagIds) => {
  return apiClient.put(API_PATHS.TAGS.UPDATE_ARTICLE_TAGS(articleId), null, {
    params: { tag_ids: tagIds },
    headers: getAuthHeaders()
  });
};

/**
 * 从文章中删除标签
 * @param {number} articleId - 文章ID
 * @param {number} tagId - 标签ID
 * @returns {Promise} 返回删除结果
 */
export const removeTagFromArticle = (articleId, tagId) => {
  return apiClient.delete(API_PATHS.TAGS.REMOVE_FROM_ARTICLE(articleId, tagId), {
    headers: getAuthHeaders()
  });
};

/**
 * 搜索标签
 * @param {string} query - 搜索关键词
 * @param {number} [limit=10] - 返回的标签数量
 * @returns {Promise} 返回搜索结果
 */
export const searchTags = (query, limit = 10) => {
  return apiClient.get(API_PATHS.SEARCH.TAGS, {
    params: {
      q: query,
      limit
    }
  });
};

export default {
  getTags,
  getTagById,
  getTagByName,
  createTag,
  updateTag,
  deleteTag,
  batchDeleteTags,
  getArticlesByTag,
  addTagToArticle,
  updateArticleTags,
  removeTagFromArticle,
  searchTags
};
